fixit/rules/chained_instance_check.py

```diff
@@ -0,0 +1,188 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import List, Optional, Set, Union
+
+import libcst as cst
+import libcst.matchers as m
+from libcst.metadata import QualifiedName, QualifiedNameProvider, QualifiedNameSource
+
+from fixit import (
+    CstContext,
+    CstLintRule,
+    InvalidTestCase as Invalid,
+    ValidTestCase as Valid,
+)
+
+
+_ISINSTANCE = QualifiedName(
+    name="builtins.isinstance", source=QualifiedNameSource.BUILTIN
+)
+
+
+class CollapseIsinstanceChecksRule(CstLintRule):
+    """
+    The built-in ``isinstance`` function, instead of a single type,
+    can take a tuple of types and check whether given target suits
+    any of them. Rather than chaining multiple ``isinstance`` calls
+    with a boolean-or operation, a single ``isinstance`` call where
+    the second argument is a tuple of all types can be used.
+    """
+
+    MESSAGE: str = (
+        "Multiple isinstance calls with the same target but "
+        + "different types can be collapsed into a single call "
+        + "with a tuple of types."
+    )
+
+    METADATA_DEPENDENCIES = (QualifiedNameProvider,)
+
+    VALID = [
+        Valid("foo() or foo()"),
+        Valid("foo(x, y) or foo(x, z)"),
+        Valid("foo(x, y) or foo(x, z) or foo(x, q)"),
+        Valid("isinstance() or isinstance()"),
+        Valid("isinstance(x) or isinstance(x)"),
+        Valid("isinstance(x, y) or isinstance(x)"),
+        Valid("isinstance(x) or isinstance(x, y)"),
+        Valid("isinstance(x, y) or isinstance(t, y)"),
+        Valid("isinstance(x, y) and isinstance(x, z)"),
+        Valid("isinstance(x, y) or isinstance(x, (z, q))"),
+        Valid("isinstance(x, (y, z)) or isinstance(x, q)"),
+        Valid("isinstance(x, a) or isinstance(y, b) or isinstance(z, c)"),
+        Valid(
+            """
+            def isinstance(x, y):
+                return None
+            if isinstance(x, y) or isinstance(x, z):
+                pass
+            """
+        ),
+    ]
+    INVALID = [
+        Invalid(
+            "isinstance(x, y) or isinstance(x, z)",
+            expected_replacement="isinstance(x, (y, z))",
+        ),
```

Comment:
The current implementation doesn't work for cases combined with other operator, e.g.
"something and isinstance(x, a) or isinstance(x, b)"
"something or isinstance(x, a) or isinstance(x, b)".


To make the implementation right, we can probably keep track of valid isinstance checks in `visit_BooleanOperation` and then try to generate the collapsed from bottom up recursion in `visit_BooleanOperation`. That way, each `visit_BooleanOperation` step tries to collapse two isintance checks and will only report the the replacement when no more collapse-able operator can be found.

Reply:
I am a bit confused on this example `something and isinstance(x, a) or isinstance(x, b)`, since it shouldn't be collapsed (precedence of AND > OR) . For this one, will try to find a solution `something or isinstance(x, a) or isinstance(x, b).`

Reply:
You're right, considering the precedence, it cannot be collapsed.

Reply:
I think I found a very simple and straight forward way of building multi-item or-expressions, but still needs much tests. Short preview;
```
        def unwrap(node: cst.BaseExpression):
            if m.matches(node, m.BooleanOperation(operator=cst.Or())):
                yield from unwrap(node.left)
                yield node.right
            else:
                yield node
        
        names = {}
        stack = tuple(unwrap(node))
...
```